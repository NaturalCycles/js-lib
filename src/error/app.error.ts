import type { ErrorData } from './error.model'

/**
 * Base class for all our (not system) errors.
 *
 * message - "technical" message. Frontend decides to show it or not.
 * data - optional "any" payload.
 * data.userMessage - if present, will be displayed to the User as is.
 *
 * Based on: https://medium.com/@xpl/javascript-deriving-from-error-properly-8d2f8f315801
 */
export class AppError<DATA_TYPE extends ErrorData = ErrorData> extends Error {
  data!: DATA_TYPE

  /**
   * cause here is normalized to be instance of Error
   */
  override cause?: Error

  constructor(message: string, data = {} as DATA_TYPE, opt?: ErrorOptions) {
    super(message)

    Object.defineProperty(this, 'name', {
      value: this.constructor.name,
      configurable: true,
    })

    Object.defineProperty(this, 'data', {
      value: data,
      writable: true,
      configurable: true,
      enumerable: false,
    })

    if (opt?.cause) {
      Object.defineProperty(this, 'cause', {
        // I'd love to do _anyToError(opt.cause) here, but it causes circular dep ;(
        value: opt.cause,
        writable: true,
        configurable: true,
        enumerable: false,
      })
    }

    // todo: check if it's needed at all!
    // if (Error.captureStackTrace) {
    //   Error.captureStackTrace(this, this.constructor)
    // } else {
    //   Object.defineProperty(this, 'stack', {
    //     value: new Error().stack, // eslint-disable-line unicorn/error-message
    //     writable: true,
    //     configurable: true,
    //   })
    // }
  }
}
