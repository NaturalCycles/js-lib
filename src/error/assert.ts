import type { ErrorData, ErrorObject } from '..'
import { _deepEquals, _isErrorObject, _stringify, AssertionError, Class } from '..'

/**
 * Evaluates the `condition` (casts it to Boolean).
 * Expects it to be truthy, otherwise throws AppError.
 *
 * Should be used NOT for "expected" / user-facing errors, but
 * vice-versa - for completely unexpected and 100% buggy "should never happen" cases.
 *
 * It'll result in http 500 on the server (cause that's the right code for "unexpected" errors).
 * Pass { backendResponseStatusCode: x } at errorData argument to override the http code (will be picked up by backend-lib).
 *
 * API is similar to Node's assert(), except:
 * 1. Throws js-lib's AppError
 * 2. Has a default message, if not provided
 *
 * Since 2024-07-10 it no longer sets `userFriendly: true` by default.
 */
export function _assert(
  condition: any, // will be evaluated as Boolean
  message?: string,
  errorData?: ErrorData,
): asserts condition {
  if (!condition) {
    throw new AssertionError(message || 'condition failed', {
      ...errorData,
    })
  }
}

/**
 * Like _assert(), but prints more helpful error message.
 * API is similar to Node's assert.equals().
 *
 * Does SHALLOW, but strict equality (===), use _assertDeepEquals() for deep equality.
 */
export function _assertEquals<T>(
  actual: any,
  expected: T,
  message?: string,
  errorData?: ErrorData,
): asserts actual is T {
  if (actual !== expected) {
    const msg =
      message ||
      ['not equal', `expected: ${_stringify(expected)}`, `got     : ${_stringify(actual)}`]
        .filter(Boolean)
        .join('\n')

    throw new AssertionError(msg, {
      ...errorData,
    })
  }
}

/**
 * Like _assert(), but prints more helpful error message.
 * API is similar to Node's assert.deepEquals().
 *
 * Does DEEP equality via _deepEquals()
 */
export function _assertDeepEquals<T>(
  actual: any,
  expected: T,
  message?: string,
  errorData?: ErrorData,
): asserts actual is T {
  if (!_deepEquals(actual, expected)) {
    const msg =
      message ||
      ['not deeply equal', `expected: ${_stringify(expected)}`, `got     : ${_stringify(actual)}`]
        .filter(Boolean)
        .join('\n')

    throw new AssertionError(msg, {
      ...errorData,
    })
  }
}

export function _assertIsError<ERR extends Error = Error>(
  err: any,
  errorClass: Class<ERR> = Error as any,
): asserts err is ERR {
  if (!(err instanceof errorClass)) {
    throw new AssertionError(
      `Expected to be instanceof ${errorClass.name}, actual typeof: ${typeof err}`,
    )
  }
}

/**
 * Asserts that passed object is indeed an Error of defined ErrorClass.
 * If yes - returns peacefully (with TypeScript assertion).
 * In not - throws (re-throws) that error up.
 */
export function _assertErrorClassOrRethrow<ERR extends Error>(
  err: any,
  errorClass: Class<ERR>,
): asserts err is ERR {
  if (!(err instanceof errorClass)) {
    // re-throw
    throw err
  }
}

export function _assertIsErrorObject<DATA_TYPE extends ErrorData = ErrorData>(
  obj: any,
): asserts obj is ErrorObject<DATA_TYPE> {
  if (!_isErrorObject(obj)) {
    throw new AssertionError(`Expected to be ErrorObject, actual typeof: ${typeof obj}`)
  }
}

export function _assertIsString(v: any, message?: string): asserts v is string {
  _assertTypeOf<string>(v, 'string', message)
}

export function _assertIsNumber(v: any, message?: string): asserts v is number {
  _assertTypeOf<number>(v, 'number', message)
}

export function _assertTypeOf<T>(v: any, expectedType: string, message?: string): asserts v is T {
  // biome-ignore lint/suspicious/useValidTypeof: ok
  if (typeof v !== expectedType) {
    const msg = message || `Expected typeof ${expectedType}, actual typeof: ${typeof v}`
    throw new AssertionError(msg)
  }
}
