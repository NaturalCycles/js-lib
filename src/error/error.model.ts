import { HttpMethod, HttpStatusCode } from '../http/http.model'
import { NumberOfMilliseconds } from '../types'

/**
 * Extendable payload object to transfer custom additional data with AppError.
 */
export interface ErrorData {
  /**
   * String constant/enum that is specific for certain type of error and never changes.
   * Allows Frontend to deal with certain type of error in a specific way.
   */
  code?: string

  /**
   * If error.message is user-friendly (can be shown to the user "as is")
   */
  userFriendly?: boolean

  /**
   * Error id in some error tracking system (e.g Sentry).
   */
  errorId?: string

  /**
   * Set to true to force reporting this error (e.g to Sentry).
   * Useful to be able to force-report e.g a 4xx error, which by default wouldn't be reported.
   * Set to false to force not-reporting it.
   */
  report?: boolean

  /**
   * Set to true or false to control error reporting on the Client-side.
   * It works as in indication/hint, not a guarantee,
   * because Client-side still need to manually check and respect this property.
   */
  reportClientSide?: boolean

  /**
   * If defined - used by SentrySharedService in backend-lib.
   * Allows to report only X% of errors of this type.
   * E.g 0.1 will report 10% of errors (and ignore the 90%)
   */
  reportRate?: number

  /**
   * Sometimes error.message gets "decorated" with extra information
   * (e.g frontend-lib adds a method, url, etc for all the errors)
   * `originalMessage` is used to preserve the original `error.message` as it came from the backend.
   */
  // originalMessage?: string
  // use .cause.message instead

  /**
   * Can be used by error-reporting tools (e.g Sentry).
   * If fingerprint is defined - it'll be used INSTEAD of default fingerprint of a tool.
   * Can be used to force-group errors that are NOT needed to be split by endpoint or calling function.
   */
  fingerprint?: string[]

  /**
   * Set when throwing an error from your backend code, to indicate desired http status code.
   * e.g throw new AppError('oj', { backendResponseStatusCode: 401 })
   */
  backendResponseStatusCode?: HttpStatusCode

  /**
   * Set to true when the error was thrown after response headers were sent.
   */
  headersSent?: boolean

  /**
   * Used in e.g http 401 error.
   */
  adminAuthRequired?: boolean

  /**
   * Used in e.g http 403 error.
   */
  adminPermissionsRequired?: string[]

  /**
   * Open-ended.
   */
  [k: string]: any
}

export interface HttpRequestErrorData extends ErrorData {
  requestUrl: string
  requestBaseUrl?: string
  requestMethod: HttpMethod
  /**
   * Conveniently combines Method and Url, respects `logWithSearchParams` (and similar) configuration.
   * E.g:
   * GET /some/url
   */
  requestSignature: string
  /**
   * Can be set to 0 if request "failed to start" or "failed to reach the server".
   */
  requestDuration: NumberOfMilliseconds
  /**
   * 0 is used for edge cases when e.g it failed to reach the server.
   */
  responseStatusCode: HttpStatusCode
}

/**
 * Sometimes there are cases when Errors come from unexpected places,
 * where err is not instanceof Error, but still looks like Error.
 */
export interface ErrorLike {
  name: string
  message: string
  stack?: string
  data?: any
  cause?: any
}

/**
 * Portable object that represents Error.
 * Has extendable generic `data` property.
 * Can be easily transformed from/to Error class via:
 * - errorObjectToAppError()
 * - appErrorToErrorObject()
 *
 * Cannot contain any properties that stock Error doesn't have, otherwise
 * it will be hard to transform it to ErrorObject.
 * Everything "extra" should go under `data`.
 *
 * Instance of AppError class should satisfy ErrorObject interface.
 */
export interface ErrorObject<DATA_TYPE extends ErrorData = ErrorData> {
  /**
   * Name of the error / error class.
   *
   * @example `Error`
   */
  // Name is removed right now, cause we should not rely on "name"
  // due to "subclassing reasons" (you lose name when you subclass)
  // It's allowed to rely on error.code and error.data.* instead
  // UPD: not anymore. Subclassing issues are fully fixed now (for es2015+ code)
  name: string

  /**
   * Error.message
   */
  message: string

  /**
   * Stacktrace of error, \n-separated, as it comes from Error.stack
   */
  stack?: string

  /**
   * Custom data to be passed with an error. Extendable.
   * It's non-optional, to save some null-checks.
   */
  data: DATA_TYPE

  cause?: ErrorObject
}

/**
 * JSON HTTP response object from the Backend that represents "Error".
 * Assumption is that if JSON response "looks like this" - it's safe to print it in a custom way.
 * Otherwise - it'll be printed in a generic way.
 */
export interface BackendErrorResponseObject<DATA_TYPE extends ErrorData = ErrorData> {
  error: ErrorObject<DATA_TYPE>
}
