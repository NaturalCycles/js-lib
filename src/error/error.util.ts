import type {
  ErrorData,
  ErrorObject,
  BackendErrorResponseObject,
  Class,
  HttpRequestErrorData,
  ErrorLike,
} from '..'
import { _jsonParseIfPossible, _stringifyAny, _truncate, _truncateMiddle } from '..'

/**
 * Useful to ensure that error in `catch (err) { ... }`
 * is indeed an Error (and not e.g `string` or `undefined`).
 * 99% of the cases it will be Error already.
 * Becomes more useful since TypeScript 4.4 made `err` of type `unknown` by default.
 *
 * Alternatively, if you're sure it's Error - you can use `_assertIsError(err)`.
 */
export function _anyToError<ERROR_TYPE extends Error = Error>(
  o: any,
  errorClass: Class<ERROR_TYPE> = Error as any,
  errorData?: ErrorData,
): ERROR_TYPE {
  let e: ERROR_TYPE

  if (o instanceof errorClass) {
    e = o
  } else {
    // If it's an instance of Error, but ErrorClass is something else (e.g AppError) - it'll be "repacked" into AppError

    const errorObject = _anyToErrorObject(o)
    e = _errorObjectToError(errorObject, errorClass)
  }

  if (errorData) {
    ;(e as any).data = {
      ...(e as any).data,
      ...errorData,
    }
  }

  return e
}

/**
 * Converts "anything" to ErrorObject.
 * Detects if it's HttpErrorResponse, HttpErrorObject, ErrorObject, Error, etc..
 * If object is Error - Error.message will be used.
 * Objects (not Errors) get converted to prettified JSON string (via `_stringifyAny`).
 */
export function _anyToErrorObject<DATA_TYPE extends ErrorData = ErrorData>(
  o: any,
  errorData?: Partial<DATA_TYPE>,
): ErrorObject<DATA_TYPE> {
  let eo: ErrorObject<DATA_TYPE>

  if (_isErrorLike(o)) {
    eo = _errorLikeToErrorObject(o)
  } else {
    o = _jsonParseIfPossible(o)

    if (_isBackendErrorResponseObject(o)) {
      eo = o.error as ErrorObject<DATA_TYPE>
    } else if (_isErrorObject(o)) {
      eo = o as ErrorObject<DATA_TYPE>
    } else if (_isErrorLike(o)) {
      eo = _errorLikeToErrorObject(o)
    } else {
      // Here we are sure it has no `data` property,
      // so, fair to return `data: {}` in the end
      // Also we're sure it includes no "error name", e.g no `Error: ...`,
      // so, fair to include `name: 'Error'`
      const message = _stringifyAny(o)

      eo = {
        name: 'Error',
        message,
        data: {} as DATA_TYPE, // empty
      }
    }
  }

  Object.assign(eo.data, errorData)
  return eo
}

export function _errorLikeToErrorObject<DATA_TYPE extends ErrorData = ErrorData>(
  e: AppError<DATA_TYPE> | Error | ErrorLike,
): ErrorObject<DATA_TYPE> {
  // If it's already an ErrorObject - just return it
  // AppError satisfies ErrorObject interface
  // Error does not satisfy (lacks `data`)
  // UPD: no, we expect a "plain object" here as an output,
  // because Error classes sometimes have non-enumerable properties (e.g data)
  if (!(e instanceof Error) && _isErrorObject(e)) {
    return e as ErrorObject<DATA_TYPE>
  }

  const obj: ErrorObject<DATA_TYPE> = {
    name: e.name,
    message: e.message,
    data: { ...(e as any).data }, // empty by default
  }

  if (e.stack) obj.stack = e.stack

  if (e.cause) {
    obj.cause = _anyToErrorObject(e.cause)
  }

  return obj
}

export function _errorObjectToError<DATA_TYPE extends ErrorData, ERROR_TYPE extends Error>(
  o: ErrorObject<DATA_TYPE>,
  errorClass: Class<ERROR_TYPE> = Error as any,
): ERROR_TYPE {
  if (o instanceof errorClass) return o

  // Here we pass constructor values assuming it's AppError or sub-class of it
  // If not - will be checked at the next step
  // We cannot check `if (errorClass instanceof AppError)`, only `err instanceof AppError`
  const { name, cause } = o
  const err = new errorClass(o.message, o.data, { name, cause })
  // name: err.name, // cannot be assigned to a readonly property like this
  // stack: o.stack, // also readonly e.g in Firefox

  if (o.stack) {
    Object.defineProperty(err, 'stack', {
      value: o.stack,
    })
  }

  if (!(err instanceof AppError)) {
    // Following actions are only needed for non-AppError-like errors

    Object.defineProperties(err, {
      name: {
        value: name,
        configurable: true,
        writable: true,
      },
      data: {
        value: o.data,
        writable: true,
        configurable: true,
        enumerable: false,
      },
      cause: {
        value: cause,
        writable: true,
        configurable: true,
        enumerable: true,
      },
    })

    Object.defineProperty(err.constructor, 'name', {
      value: name,
      configurable: true,
      writable: true,
    })
  }

  return err
}

export interface ErrorSnippetOptions {
  /**
   * Max length of the error line.
   * Snippet may have multiple lines, one original and one per `cause`.
   */
  maxLineLength?: number

  maxLines?: number
}

// These "common" error classes will not be printed as part of the Error snippet
const commonErrorClasses = new Set([
  'Error',
  'AppError',
  'AssertionError',
  'HttpRequestError',
  'JoiValidationError',
])

/**
 * Provides a short semi-user-friendly error message snippet,
 * that would allow to give a hint to the user what went wrong,
 * also to developers and CS to distinguish between different errors.
 *
 * It's not supposed to have full information about the error, just a small extract from it.
 */
export function _errorSnippet(err: any, opt: ErrorSnippetOptions = {}): string {
  const { maxLineLength = 60, maxLines = 3 } = opt
  const e = _anyToErrorObject(err)

  const lines = [errorObjectToSnippet(e)]

  let { cause } = e

  while (cause && lines.length < maxLines) {
    lines.push('Caused by ' + errorObjectToSnippet(cause))
    cause = cause.cause // insert DiCaprio Inception meme
  }

  return lines.map(line => _truncate(line, maxLineLength)).join('\n')

  function errorObjectToSnippet(e: ErrorObject): string {
    // Return snippet if it was already prepared
    if (e.data.snippet) return e.data.snippet

    // Code already serves the purpose of the snippet, so we can just return it
    if (e.data.code) return e.data.code

    return [
      !commonErrorClasses.has(e.name) && e.name,
      // replace "1+ white space characters" with a single space
      e.message.replaceAll(/\s+/gm, ' ').trim(),
    ]
      .filter(Boolean)
      .join(': ')
  }
}

export function _isBackendErrorResponseObject(o: any): o is BackendErrorResponseObject {
  return _isErrorObject(o?.error)
}

export function _isHttpRequestErrorObject(o: any): o is ErrorObject<HttpRequestErrorData> {
  return !!o && o.name === 'HttpRequestError' && typeof o.data?.requestUrl === 'string'
}

/**
 * Note: any instance of AppError is also automatically an ErrorObject
 */
export function _isErrorObject<DATA_TYPE extends ErrorData = ErrorData>(
  o: any,
): o is ErrorObject<DATA_TYPE> {
  return (
    !!o &&
    typeof o === 'object' &&
    typeof o.name === 'string' &&
    typeof o.message === 'string' &&
    typeof o.data === 'object'
  )
}

export function _isErrorLike(o: any): o is ErrorLike {
  return !!o && typeof o === 'object' && typeof o.name === 'string' && typeof o.message === 'string'
}

/**
 * Convenience function to safely add properties to Error's `data` object
 * (even if it wasn't previously existing).
 * Mutates err.
 * Returns err for convenience, so you can re-throw it directly.
 *
 * @example
 *
 * try {} catch (err) {
 *   throw _errorDataAppend(err, {
 *     httpStatusCode: 401,
 *   })
 * }
 */
export function _errorDataAppend<ERR>(err: ERR, data?: ErrorData): ERR {
  if (!data) return err
  ;(err as any).data = {
    ...(err as any).data,
    ...data,
  }

  return err
}

/**
 * Base class for all our (not system) errors.
 *
 * message - "technical" message. Frontend decides to show it or not.
 * data - optional "any" payload.
 * data.userFriendly - if present, will be displayed to the User as is.
 *
 * Based on: https://medium.com/@xpl/javascript-deriving-from-error-properly-8d2f8f315801
 */
export class AppError<DATA_TYPE extends ErrorData = ErrorData> extends Error {
  data!: DATA_TYPE

  /**
   * `cause` here is normalized to be an ErrorObject
   */
  override cause?: ErrorObject

  constructor(message: string, data = {} as DATA_TYPE, opt: AppErrorOptions = {}) {
    super(message)
    const { name = this.constructor.name, cause } = opt

    Object.defineProperties(this, {
      name: {
        value: name,
        configurable: true,
        writable: true,
      },
      data: {
        value: data,
        writable: true,
        configurable: true,
        enumerable: false,
      },
    })

    if (cause) {
      Object.defineProperty(this, 'cause', {
        value: _anyToErrorObject(cause),
        writable: true,
        configurable: true,
        enumerable: true, // unlike standard - setting it to true for "visibility"
      })
    }

    // this is to allow changing this.constuctor.name to a non-minified version
    Object.defineProperty(this.constructor, 'name', {
      value: name,
      configurable: true,
      writable: true,
    })

    // todo: check if it's needed at all!
    // if (Error.captureStackTrace) {
    //   Error.captureStackTrace(this, this.constructor)
    // } else {
    //   Object.defineProperty(this, 'stack', {
    //     value: new Error().stack, // eslint-disable-line unicorn/error-message
    //     writable: true,
    //     configurable: true,
    //   })
    // }
  }
}

/**
 * Extra options for AppError constructor.
 */
export interface AppErrorOptions {
  /**
   * Overrides Error.name and Error.constructor.name
   */
  name?: string

  /**
   * Sets Error.cause.
   * It is transformed with _anyToErrorObject()
   */
  cause?: any
}

/**
 * Error that is thrown when Http Request was made and returned an error.
 * Thrown by, for example, Fetcher.
 *
 * On the Frontend this Error class represents the error when calling the API,
 * contains all the necessary request and response information.
 *
 * On the Backend, similarly, it represents the error when calling some 3rd-party API
 * (backend-to-backend call).
 * On the Backend it often propagates all the way to the Backend error handler,
 * where it would be wrapped in BackendErrorResponseObject.
 *
 * Please note that `ErrorData.backendResponseStatusCode` is NOT exactly the same as
 * `HttpRequestErrorData.responseStatusCode`.
 * E.g 3rd-party call may return 401, but our Backend will still wrap it into an 500 error
 * (by default).
 */
export class HttpRequestError extends AppError<HttpRequestErrorData> {
  constructor(message: string, data: HttpRequestErrorData, opt?: AppErrorOptions) {
    if (data.response) {
      Object.defineProperty(data, 'response', {
        enumerable: false,
      })
    }

    super(message, data, { ...opt, name: 'HttpRequestError' })
  }

  /**
   * Cause is strictly-defined for HttpRequestError,
   * so it always has a cause.
   * (for dev convenience)
   */
  override cause!: ErrorObject
}

export class AssertionError extends AppError {
  constructor(message: string, data?: ErrorData) {
    super(message, data, { name: 'AssertionError' })
  }
}

export interface JsonParseErrorData extends ErrorData {
  /**
   * Original text that failed to get parsed.
   */
  text?: string
}

export class JsonParseError extends AppError<JsonParseErrorData> {
  constructor(data: JsonParseErrorData) {
    const message = ['Failed to parse', data.text && _truncateMiddle(data.text, 200)]
      .filter(Boolean)
      .join(': ')

    super(message, data, { name: 'JsonParseError' })
  }
}

export class TimeoutError extends AppError {
  constructor(message: string, data?: ErrorData, opt?: AppErrorOptions) {
    super(message, data, { ...opt, name: 'TimeoutError' })
  }
}

/**
 * It is thrown when Error was expected, but didn't happen
 * ("pass" happened instead).
 * "Pass" means "no error".
 */
export class UnexpectedPassError extends AppError {
  constructor() {
    super(
      'expected error was not thrown',
      {},
      {
        name: 'UnexpectedPassError',
      },
    )
  }
}
