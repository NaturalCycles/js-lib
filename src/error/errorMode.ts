/**
 * Allows to define error-controlling behaviour for batch operations.
 *
 * @default is THROW_IMMEDIATELY in most cases
 */
export enum ErrorMode {
  /**
   * Usually a default behaviour, similar as "exit early".
   */
  THROW_IMMEDIATELY = 'THROW_IMMEDIATELY',

  /**
   * Don't throw on errors, but collect them and throw as AggregateError in the end.
   */
  THROW_AGGREGATED = 'THROW_AGGREGATED',

  /**
   * Completely suppress errors, do not aggregate nor throw anything. Resilient mode.
   */
  SUPPRESS = 'SUPPRESS',
}
