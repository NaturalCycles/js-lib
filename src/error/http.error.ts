import { AppError } from './app.error'
import type { HttpErrorData } from './error.model'

/**
 * Base class for HTTP errors - errors that define HTTP error code.
 */
export class HttpError<
  DATA_TYPE extends HttpErrorData = HttpErrorData,
> extends AppError<DATA_TYPE> {
  constructor(message: string, data: DATA_TYPE) {
    super(message, data)
  }
}
