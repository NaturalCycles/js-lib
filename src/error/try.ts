import { _stringifyAny } from '../string/stringifyAny'
import type { Class } from '../typeFest'
import type { AnyFunction, ErrorDataTuple } from '../types'
import { AppError } from './app.error'
import { _assertErrorClassOrRethrow } from './assert'

/**
 * Calls a function, returns a Tuple of [error, value].
 * Allows to write shorter code that avoids `try/catch`.
 * Useful e.g. in unit tests.
 *
 * Similar to pTry, but for sync functions.
 *
 * ERR is typed as Error, not `unknown`. While unknown would be more correct,
 * according to recent TypeScript, Error gives more developer convenience.
 * In our code we NEVER throw non-errors.
 * Only possibility of non-error is in the 3rd-party library code, in these cases it
 * can be manually cast to `unknown` for extra safety.
 *
 * @example
 *
 * const [err, v] = _try(() => someFunction())
 * if (err) ...do something...
 * v // go ahead and use v
 */
export function _try<T, ERR extends Error = Error>(
  fn: () => T,
  errorClass?: Class<ERR>,
): ErrorDataTuple<T, ERR> {
  try {
    return [null, fn()]
  } catch (err) {
    if (errorClass) {
      _assertErrorClassOrRethrow(err, errorClass)
    }

    return [err as ERR, null]
  }
}

/**
 * Like _try, but for Promises.
 */
export async function pTry<T, ERR extends Error = Error>(
  promise: Promise<T>,
  errorClass?: Class<ERR>,
): Promise<ErrorDataTuple<Awaited<T>, ERR>> {
  try {
    return [null, await promise]
  } catch (err) {
    if (errorClass) {
      _assertErrorClassOrRethrow(err, errorClass)
    }
    return [err as ERR, null]
  }
}

/**
 * It is thrown when Error was expected, but didn't happen
 * ("pass" happened instead).
 * "Pass" means "no error".
 */
export class UnexpectedPassError extends AppError {
  constructor() {
    super('expected error was not thrown', {}, undefined, 'UnexpectedPassError')
  }
}

/**
 * Calls `fn`, expects is to throw, catches the expected error and returns.
 * If error was NOT thrown - throws UnexpectedPassError instead.
 *
 * If `errorClass` is passed:
 * 1. It automatically infers it's type
 * 2. It does `instanceof` check and throws if wrong Error instance was thrown.
 */
export function _expectedError<ERR = Error>(fn: AnyFunction, errorClass?: Class<ERR>): ERR {
  try {
    fn()
    // Unexpected!
    throw new UnexpectedPassError()
  } catch (err) {
    if (err instanceof UnexpectedPassError) throw err // re-throw
    if (errorClass && !(err instanceof errorClass)) {
      console.warn(
        `_expectedError expected ${errorClass.constructor.name} but got different error class`,
      )
      throw err
    }
    return err as ERR // this is expected!
  }
}

/**
 * Awaits passed `promise`, expects is to throw (reject), catches the expected error and returns.
 * If error was NOT thrown - throws UnexpectedPassError instead.
 *
 * If `errorClass` is passed:
 * 1. It automatically infers it's type
 * 2. It does `instanceof` check and throws if wrong Error instance was thrown.
 */
export async function pExpectedError<ERR = Error>(
  promise: Promise<any>,
  errorClass?: Class<ERR>,
): Promise<ERR> {
  try {
    await promise
    // Unexpected!
    throw new UnexpectedPassError()
  } catch (err) {
    if (err instanceof UnexpectedPassError) throw err // re-throw
    if (errorClass && !(err instanceof errorClass)) {
      console.warn(
        `pExpectedError expected ${errorClass.constructor.name} but got different error class`,
      )
      throw err
    }
    return err as ERR // this is expected!
  }
}

/**
 * Shortcut function to simplify error snapshot-matching in tests.
 */
export async function pExpectedErrorString<ERR = Error>(
  promise: Promise<any>,
  errorClass?: Class<ERR>,
): Promise<string> {
  const err = await pExpectedError<ERR>(promise, errorClass)
  return _stringifyAny(err)
}

/**
 * Shortcut function to simplify error snapshot-matching in tests.
 */
export function _expectedErrorString<ERR = Error>(
  fn: AnyFunction,
  errorClass?: Class<ERR>,
): string {
  const err = _expectedError<ERR>(fn, errorClass)
  return _stringifyAny(err)
}
