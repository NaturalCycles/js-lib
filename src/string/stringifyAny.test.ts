import { expectResults, mockAllKindsOfThings } from '@naturalcycles/dev-lib/dist/testing'
import { inspectAnyStringifyFn } from '@naturalcycles/nodejs-lib'
import { AppError } from '../error/app.error'
import { BackendErrorResponseObject } from '../error/error.model'
import { _errorLikeToErrorObject } from '../error/error.util'
import { pExpectedError } from '../error/try'
import { _stringifyAny, setGlobalStringifyFunction } from './stringifyAny'

test('stringifyAny default', () => {
  expectResults(v => _stringifyAny(v), mockAllKindsOfThings()).toMatchSnapshot()

  expectResults(
    v =>
      _stringifyAny(v, {
        includeErrorData: true,
      }),
    mockAllKindsOfThings(),
  ).toMatchSnapshot()
})

test('appError', () => {
  const err = new AppError('la la', {
    httpStatusCode: 409,
    userFriendly: true,
    other: 'otherValue',
  })

  expect(_stringifyAny(err)).toMatchInlineSnapshot(`"AppError: la la"`)

  expect(_stringifyAny(err, { includeErrorData: true })).toMatchInlineSnapshot(`
    "AppError: la la
    {
      "httpStatusCode": 409,
      "userFriendly": true,
      "other": "otherValue"
    }"
  `)
})

test('appError with status 0', () => {
  const err = new AppError('la la', {
    httpStatusCode: 0,
    userFriendly: true,
    other: 'otherValue',
  })

  expect(_stringifyAny(err)).toMatchInlineSnapshot(`"AppError: la la"`)
})

test('backendErrorResponse', () => {
  const err = new AppError('la la\nsecond line', {
    httpStatusCode: 409,
    userFriendly: true,
    other: 'otherValue',
  })

  const resp: BackendErrorResponseObject = {
    error: _errorLikeToErrorObject(err),
  }
  expect(resp.error.name).toBe('AppError')

  expect(_stringifyAny(resp)).toMatchInlineSnapshot(`
    "AppError: la la
    second line"
  `)

  // this tests "duplicated line" bug
  // expect(_stringifyAny(resp, { includeErrorStack: true })).toMatchInlineSnapshot()
})

test('error with cause', () => {
  const err = new Error('err1', {
    cause: new AppError(
      'http_error1',
      {
        httpStatusCode: 400,
      },
      {
        cause: {
          name: 'SomeError',
          message: 'sub-cause',
          data: {},
        },
      },
    ),
  })

  expect(_stringifyAny(err)).toMatchInlineSnapshot(`
    "Error: err1
    Caused by: AppError: http_error1
    Caused by: SomeError: sub-cause"
  `)
})

test('AggregateError', async () => {
  const err = await pExpectedError(
    Promise.any([
      new Promise((_, reject) => reject(new Error('err1'))),
      new Promise((_, reject) => reject(new Error('err2'))),
    ]),
    AggregateError,
  )

  expect(_stringifyAny(err)).toMatchInlineSnapshot(`
    "AggregateError: All promises were rejected
    2 error(s):
    1. Error: err1
    2. Error: err2"
  `)
})

const obj = {
  a: 'a',
  b: { c: 'c' },
}

test('simple object', () => {
  expect(_stringifyAny(obj)).toMatchInlineSnapshot(`
    "{
      "a": "a",
      "b": {
        "c": "c"
      }
    }"
  `)
})

test('setGlobalStringifyFunction', () => {
  setGlobalStringifyFunction(inspectAnyStringifyFn)

  expect(_stringifyAny(obj)).toMatchInlineSnapshot(`"{ a: 'a', b: { c: 'c' } }"`)
})
