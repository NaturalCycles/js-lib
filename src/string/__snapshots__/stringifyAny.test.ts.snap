// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyAny default 1`] = `
Map {
  undefined => "undefined",
  null => "null",
  true => "true",
  false => "false",
  NaN => "null",
  Infinity => "null",
  "" => "empty_string",
  " " => "empty_string",
  "ho ho ho" => "ho ho ho",
  "{" => "{",
  "{"a": "a1"}" => "{
  "a": "a1"
}",
  "[5, null, "s"]" => "[
  5,
  null,
  "s"
]",
  [
    5,
    null,
    "s",
  ] => "[
  5,
  null,
  "s"
]",
  15 => "15",
  [
    1,
    2,
    3,
  ] => "[
  1,
  2,
  3
]",
  {
    "a": "a1",
    "b": "b1",
    "c": {
      "d": 25,
      "e": undefined,
    },
  } => "{
  "a": "a1",
  "b": "b1",
  "c": {
    "d": 25
  }
}",
  {
    "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference",
    "message": "Reference already exists",
  } => "{
  "message": "Reference already exists",
  "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference"
}",
  [Function] => "function",
  /i am regex, who are you\\?/ => "{}",
  Map {
    "a" => "b",
  } => "{}",
  Set {
    "a",
    "b",
    "c",
  } => "{}",
  [Error: some err msg] => "Error: some err msg",
  [Error: Error with code] => "Error: Error with code
code: auth/forbidden",
  [AppError: app error] => "AppError: app error",
  Symbol(a_am_symbol) => "Symbol(a_am_symbol)",
  1984-06-21T00:00:00.000Z => ""1984-06-21T00:00:00.000Z"",
}
`;

exports[`stringifyAny default 2`] = `
Map {
  undefined => "undefined",
  null => "null",
  true => "true",
  false => "false",
  NaN => "null",
  Infinity => "null",
  "" => "empty_string",
  " " => "empty_string",
  "ho ho ho" => "ho ho ho",
  "{" => "{",
  "{"a": "a1"}" => "{
  "a": "a1"
}",
  "[5, null, "s"]" => "[
  5,
  null,
  "s"
]",
  [
    5,
    null,
    "s",
  ] => "[
  5,
  null,
  "s"
]",
  15 => "15",
  [
    1,
    2,
    3,
  ] => "[
  1,
  2,
  3
]",
  {
    "a": "a1",
    "b": "b1",
    "c": {
      "d": 25,
      "e": undefined,
    },
  } => "{
  "a": "a1",
  "b": "b1",
  "c": {
    "d": 25
  }
}",
  {
    "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference",
    "message": "Reference already exists",
  } => "{
  "message": "Reference already exists",
  "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference"
}",
  [Function] => "function",
  /i am regex, who are you\\?/ => "{}",
  Map {
    "a" => "b",
  } => "{}",
  Set {
    "a",
    "b",
    "c",
  } => "{}",
  [Error: some err msg] => "Error: some err msg",
  [Error: Error with code] => "Error: Error with code
code: auth/forbidden",
  [AppError: app error] => "AppError: app error
{
  "k1": "v1"
}",
  Symbol(a_am_symbol) => "Symbol(a_am_symbol)",
  1984-06-21T00:00:00.000Z => ""1984-06-21T00:00:00.000Z"",
}
`;
